llama.cpp

```diff
@@ -7377,18 +7377,15 @@ void llama_sample_min_p(struct llama_context * ctx, llama_token_data_array * can
     std::vector<llama_token_data> filtered_candidates;
     filtered_candidates.reserve(candidates->size);  // Reserve to avoid multiple reallocations
 
-    size_t kept_count = 0;  // Counter for how many tokens are kept
-
     for (size_t i = 0; i < candidates->size; ++i) {
-        // If a token's probability is above the threshold, we keep it.
-        if (candidates->data[i].p >= multiplied_min_p) {
+        // If a token's probability is above the threshold or if we haven't kept enough tokens yet
+        if (candidates->data[i].p >= multiplied_min_p || filtered_candidates.size() < min_keep) {
```

Comment:
(Yes the whitespace thing seems good now.)

This actually looks weird because the list is already sorted, right?

So firstly, the first time the condition at 7382 fails, you can just `break`. Also at that point there's no reason for to check if you met the min keep threshold. You would have kept iterating the loop if you could have, pulling in the top entries to satisfy `min_keep`. Right?

Hmm, actually this is all more complicated than it needs to be. I believe this is all you need:

```c++
void llama_sample_min_p(struct llama_context * ctx, llama_token_data_array * candidates, float p, size_t min_keep) {
    float base_min_p = p;  // This will hold the base minimum probability value
    float multiplied_min_p;  // This will hold the adjusted minimum probability threshold

    // Ensure the probabilities are calculated.
    llama_sample_softmax(ctx, candidates);

    // Calculate the multiplication factor based on the highest scoring token.
    float multiplication_factor = candidates->data[0].p;

    // Calculate the minimum percentage requirement.
    multiplied_min_p = base_min_p * multiplication_factor;

    for (size_t kept_count = 0; kept_count < candidates->size; kept_count++) {
        if (kept_count >= min_keep && candidate->data[kept_count].p < multiplied_min_p) {
            break;
        }
    }
    candidates->size = kept_count;
    // That's it, you're done.
}
```

Reply:
kept_count is never declared here, and there's a typo where candidate->data should be candidates->data if I'm not wrong.

Reply:
Yeah, I also tried rewriting it this way. I see no reason why it shouldn't work here if it works for top_p.

Reply:
No problem. Think I edited the out issues in that example code now (though I didn't test or even try to compile it). You should look at it carefully as I'm currently falling asleep.

Reply:
I like @KerfuffleV2 solution.
@kalomaze Note that min_p sampler should be disabled when parameter p is equal to 1.0

Reply:
Actually I guess that first line should be: `if (candidates->size < 1 || p == 0.0f) return;`

That doesn't honor `min_keep` but neither does the top-p sampler, so...